File: Tasker.Core/BL/Managers/TaskManager.cs
using System;
using System.Collections.Generic;

using Tasker.Core.BL.Contracts;
using Tasker.Core.DAL.Contracts;
using Tasker.Core.DAL.Entities;

namespace Tasker.Core.BL.Managers
{
    public class TaskManager : ITaskManager
    {
        private IRepository<Task> _taskRepository;
        private IRepository<Project> _projectRepository;

        public TaskManager(IUnitOfWork unitOfWork)
        {
            _taskRepository = unitOfWork.TasksRepository;
            _projectRepository = unitOfWork.ProjectsRepository;
        }

        public Task Get(int id)
        {
            return _taskRepository.GetById(id);
        }

        public List<Task> GetAll()
        {
            var tasks = _taskRepository.GetAll();
            var list = new List<Task>();
            if (tasks != null)
            {
                list.AddRange(tasks);
            }
            return list;
        }

        public List<Task> GetAllOpen()
        {
            return GetAll().FindAll(t => !t.IsCompleted);
        }

        public List<Task> GetAllCompleted()
        {
            return GetAll().FindAll(t => t.IsCompleted);
        }

        public List<Task> GetProjectTasks(int projectId)
        {
            var tasks = _taskRepository.Find(x => x.ProjectID == projectId);
            var list = new List<Task>();
            if (tasks != null)
            {
                list.AddRange(tasks);
            }
            return list;
        }

        public List<Task> GetProjectOpenTasks(int projectId)
        {
            return GetProjectTasks(projectId).FindAll(t => !t.IsCompleted);
        }

        public List<Task> GetProjectCompletedTasks(int projectId)
        {
            return GetProjectTasks(projectId).FindAll(t => t.IsCompleted);
        }

        public List<Task> GetWhere(Predicate<Task> predicate)
        {
            return GetAll().FindAll(predicate);
        }

        public List<Project> GetProjects()
        {
            var projects = _projectRepository.GetAll();
            var list = new List<Project>();
            if (projects != null)
            {
                list.AddRange(projects);
            }
            return list;
        }

        public List<Task> GetForToday()
        {
            return GetAllOpen().FindAll(t => t.DueDate < DateTime.Today.AddDays(1));
        }

        public List<Task> GetForTomorrow()
        {
            return GetAllOpen().FindAll(t => t.DueDate < DateTime.Today.AddDays(2) && t.DueDate >= DateTime.Today.AddDays(1));
        }

        public List<Task> GetForNextWeek()
        {
            return GetAllOpen().FindAll(t => t.DueDate < DateTime.Today.AddDays(8));
        }

        public int SaveItem(Task item)
        {
            if (item.ID != 0)
            {
                var previousVersion = _taskRepository.GetById(item.ID);
                if (item.ProjectID != previousVersion.ProjectID)
                {
                    if(previousVersion.ProjectID != 0)
                    {
                        var previousProject = _projectRepository.GetById(previousVersion.ProjectID);
                        if (previousVersion.IsCompleted)
                        {
                            previousProject.CountOfCompletedTasks--;
                        }
                        else
                        {
                            previousProject.CountOfOpenTasks--;
                        }
                        _projectRepository.Save(previousProject);
                    }
                    if (item.ProjectID != 0)
                    {
                        var project = _projectRepository.GetById(item.ProjectID);
                        if (!item.IsCompleted)
                        {
                            project.CountOfOpenTasks++;
                        }
                        else
                        {
                            project.CountOfCompletedTasks++;
                        }

                        _projectRepository.Save(project);
                    }
                }
            }
            else if (item.ProjectID != 0 )
            {
                var project = _projectRepository.GetById(item.ProjectID);
                if (!item.IsCompleted)
                {
                    project.CountOfOpenTasks++;
                }
                else
                {
                    project.CountOfCompletedTasks++;
                }
               
                _projectRepository.Save(project);
            }
            return _taskRepository.Save(item);
        }

        public void ChangeStatus(int id)
        {
            var task = Get(id);
            ChangeStatus(task);
        }

        public void ChangeStatus(Task task)
        {
            if (task.ProjectID != 0)
            {
                var project = _projectRepository.GetById(task.ProjectID);
                if (!task.IsCompleted)
                {
                    
                    project.CountOfCompletedTasks++;
                    project.CountOfOpenTasks--;
                }
                else
                {
                    project.CountOfCompletedTasks--;
                    project.CountOfOpenTasks++;
                }
                _projectRepository.Save(project);
            }
            task.CompletedDate = DateTime.Now;
            task.IsCompleted = !task.IsCompleted;
            _taskRepository.Save(task);
        }

        public int Delete(int id)
        {
            var task = Get(id);
            if (task.ProjectID != 0)
            {
                var project = _projectRepository.GetById(task.ProjectID);

                if (task.IsCompleted)
                {
                    project.CountOfCompletedTasks--;
                }
                else
                {
                    project.CountOfOpenTasks--;
                }

                _projectRepository.Save(project);
            }

            return _taskRepository.Delete(id);
        }

        public int Delete(Task item)
        {
            return Delete(item.ID);
        }

        public int DeleteGroup(IList<Task> group)
        {
            int count = 0;
            foreach (var task in group)
            {
                Delete(task.ID);
                count++;
            }
            return count;
        }


    }
}

File: Native/core/src/main/java/com/cybernyanta/core/manager/TaskManager.java
package com.cybernyanta.core.manager;

import com.cybernyanta.core.database.FirebaseArrayList;
import com.cybernyanta.core.database.OnChangedListener;
import com.cybernyanta.core.model.Project;
import com.cybernyanta.core.model.Task;

import java.util.ArrayList;
import java.util.Collections;
import java.util.Date;
import java.util.List;

/**
 * Created by evgeniy.siyanko on 04.01.2017.
 */

public class TaskManager{

    private FirebaseArrayList<Task> datasource;

    public TaskManager(FirebaseArrayList<Task> datasource) {
        this.datasource = datasource;
    }

    public void addOnChangedListener(OnChangedListener listener) {
        datasource.addOnChangedListener(listener);
    }

    public void cleanup() {
        datasource.cleanup();
    }

   /* public List<Task> GetProjectTasks(String projectId){
    }

    public List<Task> GetProjectOpenTasks(String projectId){
    }

    public List<Task> GetProjectCompletedTasks(String projectId){

    }*/

    public List<Task> GetAllOpen(){
        List<Task> result = new ArrayList<>();
        for(Task task:datasource){
            if(!task.isCompleted()){
                result.add(task);
            }
        }
        return  result;
    }

    public List<Task> GetAllCompleted(){
        List<Task> result = new ArrayList<>();
        for(Task task:datasource){
            if(task.isCompleted()){
                result.add(task);
            }
        }
        return result;
    }

/*    public List<Task> GetWhere(Predicate<Task> predicate){

    }*/

   /* public List<Project> GetProjects(){

    }*/

    public List<Task> GetForToday(){
        List<Task> result = new ArrayList<>();
        for(Task task:datasource){
            if(task.getDueDate().compareTo(new Date())==1){
                result.add(task);

            }
        }
        return  result;
    }

    public List<Task> GetForTomorrow(){
        return null;
    }

    public List<Task> GetForNextWeek(){
        return null;
    }

    public void ChangeStatus(int id){

    }

    public void ChangeStatus(Task task){

    }
}
